v4/export/dump.go

```diff
@@ -3,112 +3,100 @@ package export
 	if err != nil {
-		return nil, ServerInfoUnknown, withStack(err)
+		return withStack(err)
 	}
 	defer pool.Close()
 
-	if conf.Database == "" {
-		databases, err = ShowDatabases(pool)
-		if err != nil {
-			return nil, ServerInfoUnknown, withStack(err)
-		}
-	} else {
-		databases = strings.Split(conf.Database, ",")
+	conf.ServerInfo, err = detectServerInfo(pool)
+	if err != nil {
+		return err
 	}
 
-	versionStr, err := SelectVersion(pool)
+	databases, err := prepareDumpingDatabases(conf, pool)
 	if err != nil {
-		return databases, ServerInfoUnknown, withStack(err)
+		return err
 	}
-	info = ParseServerInfo(versionStr)
-	return
-}
 
-func dumpDatabase(ctx context.Context, conf *Config, dbName string, writer Writer) error {
-	dsn := conf.getDSN(dbName)
-	db, err := sql.Open("mysql", dsn)
+	conf.Tables, err = listAllTables(pool, databases)
 	if err != nil {
-		return withStack(err)
+		return err
 	}
-	defer db.Close()
 
-	createDatabaseSQL, err := ShowCreateDatabase(db, dbName)
+	conCtrl := ConsistencyController{}
+	err = conCtrl.Setup(conf, ConsistencyNone)
 	if err != nil {
 		return err
 	}
-	if err := writer.WriteDatabaseMeta(ctx, dbName, createDatabaseSQL); err != nil {
+
+	fsWriter, err := NewSimpleWriter(conf)
+	if err != nil {
 		return err
 	}
 
-	tables, err := ShowTables(db)
-	if err != nil {
+	if err = dumpDatabases(context.Background(), conf, pool, fsWriter); err != nil {
 		return err
 	}
+	return conCtrl.TearDown()
+}
 
-	rateLimit := newRateLimit(conf.Threads)
-	var wg sync.WaitGroup
-	wg.Add(len(tables))
-	res := make([]error, len(tables))
-	for i, table := range tables {
-		go func(ith int, table string, wg *sync.WaitGroup, res []error) {
-			defer wg.Done()
-			createTableSQL, err := ShowCreateTable(db, dbName, table)
-			if err != nil {
-				res[ith] = err
-				return
-			}
-			if err := writer.WriteTableMeta(ctx, dbName, table, createTableSQL); err != nil {
-				res[ith] = err
-				return
-			}
+func dumpDatabases(ctx context.Context, conf *Config, db *sql.DB, writer Writer) error {
+	allTables := conf.Tables
+	for dbName, tables := range allTables {
+		createDatabaseSQL, err := ShowCreateDatabase(db, dbName)
+		if err != nil {
+			return err
+		}
+		if err := writer.WriteDatabaseMeta(ctx, dbName, createDatabaseSQL); err != nil {
+			return err
+		}
 
-			rateLimit.getToken()
-			tableIR, err := SelectAllFromTable(conf, db, dbName, table)
-			defer rateLimit.putToken()
+		rateLimit := newRateLimit(conf.Threads)
+		var wg sync.WaitGroup
+		wg.Add(len(tables))
+		res := make([]error, len(tables))
+		for i, table := range tables {
+			go func(ith int, table string, wg *sync.WaitGroup, res []error) {
+				defer wg.Done()
+				rateLimit.getToken()
+				defer rateLimit.putToken()
+				res[ith] = dumpTable(ctx, conf, db, dbName, table, writer)
+			}(i, table, &wg, res)
+		}
+		wg.Wait()
+		for _, err := range res {
 			if err != nil {
-				res[ith] = err
-				return
+				return err
 			}
+		}
```

Comment:
Could we use [errgroup](https://godoc.org/golang.org/x/sync/errgroup) for this?

BTW this will dump all tables even if one of them failed (late exit), while `errgroup` will stop as soon as one of the table failed (early exit). Not sure if which is the desired behavior.